Implement clearing behavior for AIP-52

dstandish opened the pull request:
When clearing a work task, it's setups and teardowns should be cleared.

It's setups are those setups which are upstream and which have teardowns that are downstream of the work task (or no teardowns).

Need to make sure to add test for the no teardown case.

Closes: #29329

ephraimbuddy commented on airflow/models/dag.py:
```diff
@@ -2207,6 +2207,9 @@ def partial_subset(
                 also_include.extend(t.get_flat_relatives(upstream=True))
+            else:
+                upstream_setup_tasks = t.get_flat_relatives(upstream=True, setup_only=True)
+                also_include.extend(upstream_setup_tasks)
```
From the AIP, clearing a work task automatically clears the task's setup/teardown tasks. If so, I think it won't work here due to the 'else' statement?

dstandish replied:
which task do you think it's missing? 
i did add tests so pretty sure it's working properly.

i bet that the confusing bit here is that, there is logic, in `get_flat_relatives` such that, whenever a setup is returned, its teardown (if one exists) is returned

so, when you're going upstream, you don't need to do anything different.

but when you're going downstream, you need to add upstream setups (and their towndows)

so, with downstream, the teardowns are added on line 2207, since they are always downstream.  

and then on 2211, it's just saying "if downstream and not upstream" then we must _add_ the upstream setups.

now, you could argue that, it's not obvious from the signature / name of get_flat_relatives that when setups are returned that their teardowns are also returned.   and if that's your thought i welcome any suggestion.

ephraimbuddy replied:
I think I didn't understand the code well. Your explanation makes sense. I was thinking that if `include_upstream` is selected by the user then the else part will not run but from your explanation it's handled in get_flat_relatives

dstandish replied:
yeah get_flat_relatives was updated so that.... whenever a setup is returned, it always adds the teardown.
now, this assumes that the only usage of it is clearing behavior.  i should probably verify that.

ephraimbuddy reviewed (commented):
I'm thinking of a slightly different approach here.
How about implementing everything in the partial_subset function 
e.g given a task to clear
1. Check the root tasks in that task_group and find the setup tasks
2. Check the leaves in that task_group and find teardown task where one of the upstream is 1 above
3. then include 1 and 2.
WDYT?

ephraimbuddy commented on airflow/models/abstractoperator.py:
```diff
@@ -158,33 +158,57 @@ def get_flat_relative_ids(
+                    is_relevant_setup = not task.downstream_task_ids.isdisjoint(downstream_teardowns)
+                    if setup_only and not is_relevant_setup:
+                        continue
+                    else:
```
It seems like the `else` is not needed.

dstandish committed:
Implement clearing behavior for AIP-52

When clearing a work task, it's setups and teardowns should be cleared.

It's setups are those setups which are upstream and which have downstream teardowns (or no teardowns).

dstandish committed:
fix precedence in setup teardown example dag classic

dstandish committed:
fix mypy

dstandish committed:
fix dag structure

dstandish committed:
fix tests

dstandish committed:
remove unnec else

dstandish committed:
attrs now public

dstandish committed:
add comments

dstandish committed:
Merge branch 'main' into implement-clearing-behavior

dstandish committed:
duplicate after merge

dstandish committed:
kwargs only

dstandish committed:
docstring

dstandish committed:
add tests

dstandish committed:
Merge branch 'main' into implement-clearing-behavior

dstandish committed:
Merge branch 'main' into implement-clearing-behavior

dstandish committed:
add more tests

dstandish committed:
--wip-- [skip ci]

dstandish committed:
simplify functions and updte tests

dstandish committed:
Merge branch 'main' into implement-clearing-behavior

dstandish committed:
remove wip

dstandish committed:
revert flat relatives change

dstandish committed:
move

dstandish committed:
fixup! move

dstandish committed:
fixup! fixup! move

dstandish committed:
use generators

dstandish committed:
simplify tests

dstandish committed:
simplify tests

dstandish committed:
move to dag

dstandish committed:
tests

dstandish committed:
tests

jedcunningham commented on tests/models/test_dag.py:
```diff
@@ -3524,3 +3519,259 @@ def test_create_dagrun_disallow_manual_to_use_automated_run_id(run_id_type: DagR
+        s1 >> w1 >> w2 >> w3
+
+        # there is no teardown downstream of w2, so we assume w2 does not need s1
+        assert set(w2.get_upstreams_only_setups_and_teardowns()) == set()
```
I feel like we should check that w1 does consider s1 a setup here before we add any other relationships.

dstandish replied:
Yeah this was a point of ambiguity that I was meaning to bring up.  For now the behavior is you need to throw on an "empty" teardown in order to make the setup's scope well-defined.  @potiuk seemed to like that behavior.  the alternative is if a setup doesn't have a teardown then all its downstream tasks are assumed to be in its scope.   and that's definitely feasible and i don't really mind either way.  lmkyt

jedcunningham replied:
Works for me, thanks for the explanation.

dstandish replied:
i made the change already, so now the downstreams of a teardownless setup are all in the scope of that setup, so no empty teardown required. what do you think is best behavior?

jedcunningham replied:
Sorry, I follow now. I don't have a strong preference either way to be honest. Both are clunky in their own way. But if we go back the original, we should check that w1 considers s1 a setup.

dstandish replied:
original is, if we don't add empty teardown, then w1 would *not* consider s1 a setup unless s1 >> t1 and t1 downstream of w1.
so i think what we'd want to check in that scenario is that w1 does *not* consider s1 a setup

jedcunningham replied:
Daniel and I had a chat and decided to go with the empty teardown approach for defining scope.

jedcunningham commented on tests/models/test_dag.py:
```diff
@@ -3524,3 +3519,259 @@ def test_create_dagrun_disallow_manual_to_use_automated_run_id(run_id_type: DagR
+        # (this is how we signal "this is the teardown for this setup")
+        # so still, we don't regard s1 as a setup for w2
+        assert set(w2.get_upstreams_only_setups_and_teardowns()) == set()
+        assert self.cleared_downstream(w2) == {w2, w3, t1}
```
And probably worth checking that w1 now has both s1 and t1.

uranusjr commented on airflow/models/abstractoperator.py:
```diff
@@ -188,6 +188,33 @@ def get_flat_relatives(self, upstream: bool = False) -> Collection[Operator]:
         return [dag.task_dict[task_id] for task_id in self.get_flat_relative_ids(upstream=upstream)]
 
+    def get_upstreams_follow_setups(self) -> Iterable[Operator]:
+        """All upstreams and, for each upstream setup, its respective teardowns."""
```
Let’s mark these functions with `:meta private:`

dstandish replied:
I thought about it but the whole class is marked private so is it necessary?  If it is necessary, what is the meaning / import of the class being private?

uranusjr replied:
Hmm, OK. I thought since this class is subclassed by public ones (e.g. BaseOperator), this would have shown up there if the function is not also marked. But I just checked and it’s not the case (for example `get_flat_relatives` is not available in docs), so you’re right a mark is not needed.

dstandish replied:
Yeah.... the problem with `:meta private:` is it has no effect on IDE behavior.  So while it's technically private, users would have a tough time seeing that.

So. leading underscore is much better, from that perspective.

However, if we use leading underscore, then, wherever we use it, it looks like a bad usage.  Because we're trying to indicate private w.r.t. user, not private w.r.t. the class (which is what underscore means).

and the IDE can't distinguish between user and internal. Wish there was a better solution.  But for now seems meta private is best we can do.

uranusjr replied:
I think Python mostly just assumes users are responsible and shouldn’t use any undocumented things by default (and Sphinx’s behaviour is good in that regard). Sure modern IDEs make the line more blurry but IMO we just need to maintain the traditional “private until documented” approach.

dstandish committed:
assume all down tasks in setup scope if it has no teardowns

dstandish committed:
Merge branch 'main' into implement-clearing-behavior

uranusjr commented on airflow/models/abstractoperator.py:
```diff
@@ -188,6 +188,44 @@ def get_flat_relatives(self, upstream: bool = False) -> Collection[Operator]:
+            if not task.downstream_task_ids.isdisjoint(downstream_teardown_ids):
+                is_relevant = True
+            # setup has no teardowns
+            elif not [x for x in task.downstream_list if x.is_teardown]:
```
```suggestion
            elif not any(x.is_teardown for x in task.downstream_list):
```

(Not sure if I got this right, the point is to use `any` or `all`)

dstandish replied:
i think that's tright

jedcunningham commented on tests/models/test_dag.py:
```diff
@@ -3599,34 +3599,49 @@ def test_get_flat_relative_ids_with_setup(self):
         assert self.cleared_downstream(w2) == {s1, w2, w3, t1}
+        assert self.cleared_downstream(w3) == {w3}
+
+        # and let's check clearing upstream just cus
```
```suggestion
        # and let's check clearing upstream
```
:)

dstandish committed:
Update tests/models/test_dag.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update airflow/models/abstractoperator.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

dstandish committed:
Merge branch 'main' into implement-clearing-behavior

dstandish committed:
revert orphaned setup change

dstandish committed:
Merge remote-tracking branch 'origin/implement-clearing-behavior' into implement-clearing-behavior

dstandish committed:
nothing

dstandish committed:
blank lines

dstandish committed:
add validation for setup without teardown
